lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1279,6 +1282,248 @@ SmallVector<DictionaryAttr> InstanceOp::portAttributes() {
+           << "cannot reference the entry-point component: '" << entryPointName
+           << "'.";
+
+  // Verify the instance result ports with those of its referenced component.
+  SmallVector<hw::PortInfo> primitivePorts = referencedPrimitive.getAllPorts();
+  size_t numPorts = primitivePorts.size();
+
+  size_t numResults = instance.getNumResults();
+  if (numResults != numPorts)
+    return instance.emitOpError()
+           << "has a wrong number of results; expected: " << numPorts
+           << " but got " << numResults;
+
+  for (size_t i = 0; i != numResults; ++i) {
+    auto resultType = instance.getResult(i).getType();
+    auto expectedType = primitivePorts[i].type;
+    auto replacedType = hw::evaluateParametricType(instance.getLoc(), instance.parametersAttr(), expectedType);
+    if (resultType == replacedType)
+      continue;
+    return instance.emitOpError()
+           << "result type for " << primitivePorts[i].name << " must be "
+           << expectedType << ", but got " << resultType;
+  }
+  return success();
+}
+
+LogicalResult PrimitiveOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  Operation *op = *this;
+  auto program = op->getParentOfType<ProgramOp>();
+  Operation *referencedPrimitive =
+      symbolTable.lookupNearestSymbolFrom(program, primitiveNameAttr());
+  if (referencedPrimitive == nullptr)
+    return emitError() << "referencing component: '" << primitiveName()
+                       << "', which does not exist.";
+
+  Operation *shadowedPrimitiveName =
+      symbolTable.lookupNearestSymbolFrom(program, sym_nameAttr());
+  if (shadowedPrimitiveName != nullptr)
+    return emitError() << "instance symbol: '" << instanceName()
+                       << "' is already a symbol for another component.";
+
+  // Verify the referenced component is not instantiating itself.
+  auto parentPrimitive = op->getParentOfType<hw::HWModuleExternOp>();
+  if (parentPrimitive == referencedPrimitive)
+    return emitError() << "recursive instantiation of its parent component: '"
+                       << primitiveName() << "'";
+
+  assert(isa<hw::HWModuleExternOp>(referencedPrimitive) && "Should be a PrimitiveOp.");
+  return verifyPrimitiveOpType(*this, cast<hw::HWModuleExternOp>(referencedPrimitive));
+}
+
+/// Provide meaningful names to the result values of an PrimitiveOp.
+void PrimitiveOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
+  getCellAsmResultNames(setNameFn, *this, this->portNames());
+}
+
+SmallVector<StringRef> PrimitiveOp::portNames() {
+  SmallVector<StringRef> portNames;
+  for (hw::PortInfo port : getReferencedPrimitive().getAllPorts())
+    portNames.push_back(port.name.getValue());
+
+  return portNames;
+}
+
+Direction convertHWDirectionToCalyx(hw::PortDirection direction) {
+  switch (direction) {
+    case hw::PortDirection::INPUT:
+      return Direction::Input;
+    case hw::PortDirection::OUTPUT:
+      return Direction::Output;
+    default:
```

Comment:
Why is there a default case? Why not use all the hw::PortDirections?

Reply:
calyx::Direction does not support InOut ports so I'm not sure how to handle the hw::PortDirection::INOUT case. Any suggestions here?

Reply:
Probably some assertion/compiler failure then. I tend to avoid default for enumeration switches because it may lead to buggy behavior when a new enumeration is added.